plugin/pkg/auth/authenticator/password/keystone/keystone.go

```diff
@@ -45,18 +50,74 @@ func (keystoneAuthenticator *KeystoneAuthenticator) AuthenticatePassword(usernam
 		glog.Info("Failed: Starting openstack authenticate client")
 		return nil, false, errors.New("Failed to authenticate")
 	}
-
+//FIXME user_id, not user_name...
 	return &user.DefaultInfo{Name: username}, true, nil
 }
 
-// NewKeystoneAuthenticator returns a password authenticator that validates credentials using openstack keystone
-func NewKeystoneAuthenticator(authURL string) (*KeystoneAuthenticator, error) {
-	if !strings.HasPrefix(authURL, "https") {
-		return nil, errors.New("Auth URL should be secure and start with https")
+func (keystoneAuthenticator *KeystoneAuthenticator) AuthenticateToken(value string) (user.Info, bool, error) {
+//FIXME kfox1111: To enhance performance, cache admin token/client.
+	cfg, provider, err := loadConfig(keystoneAuthenticator.configPath)
+	if err != nil {
+		return nil, false, errors.New("Internal error getting the Keystone provider")
+	}
+
+	client, err := openstack.NewIdentityAdminV3(provider, gophercloud.EndpointOpts{
+		Region: cfg.Global.Region,
+	})
+	if err != nil {
+		return nil, false, err
+	}
+	response, err := client.Request("GET", client.ServiceURL("auth", "tokens"), gophercloud.RequestOpts{
+	        MoreHeaders: map[string]string{"X-Subject-Token": value},
+	        OkCodes:     []int{200},
+	})
+	if err != nil {
+		return nil, false, err
+	}
+	defer response.Body.Close()
+	bodyBytes, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		glog.V(3).Infof("Cannot get HTTP response body from keystone token validate: %v", err)
+		return nil, false, err
+	}
+	obj := struct{ Token struct { User struct { Id string }}}{}
+	err = json.Unmarshal(bodyBytes, &obj)
+	if err != nil {
+		return nil, false, err
+	}
+	return &user.DefaultInfo{Name: obj.Token.User.Id}, response.StatusCode == 200, nil
```

Comment:
Keystone doesn't quite work that way.

Maybe a short Keystone 101 is in order.

Keystone is at its heart, an authz mechanism, and somewhat an authn mechanism. They have stated that they would like to mostly get out of the authn business too. Its hard. :)

Keystone is about centralizing a bunch of the cloud authz stuff in one place so its much easier to manage for operators, and then authn/authz code doesn't have to be scattered around the various cloud services, making it much easier to write services and secure them. They only need to talk to Keystone.

Keystone provides the following abstractions to assist services with getting their part of authz done in this kind of environment:

Domains - A container where users, groups, and projects are housed. Each is owned by a single domain, and names are unique only within a single domain. Each has a unique identifier that is global. Authentication plugins can be associated with a domain so that the users housed by the domain can have one type of authentication associated with it. Services usually don't have to care about domains.

Project (aka tenant). The base unit of ownership/tenancy in Keystone. OpenStack services that create resources on behalf of Users by request are associated with a Project, not with Users. Users can come and go on a Project but the Project, and its resources stays.

Group. Just a list of users. Group members have to be in the same domain as the Group.

Role. A tag. Unit of authorization.

Users are associated with a Project by a Role or a User to a Group, and a Group to a Project by a Role. Services check to see if an authenticated User has a Role on a Project to determine if the User is authorized to perform some action to resources associated with the Project. Create, delete, list, etc.

In the User/Group <- Role -> Project association, Users/Groups and Projects do not have to be in the same domain. In the case of Trusts, they often aren't.

Trust. An association between one User (Trustor) and another User (Trustee). This allows Roles on Projects to be delegated from a Human User to a Service Account User so that the Service Account can be used to perform actions on behalf of the Human without the Human's interaction or credentials. The trustee user has its own credentials.

Token, an abstract short term bearer token string that, after a User authenticates with Keystone then gives it to a Service, and that gives Services a way to validate the User is already authenticated (is token valid or not) and get authz information about:
 * What Project the User is currently acting on behalf of.
 * What Roles the user has access to on that Project, that its acting on.

So, in a multitenant supporting Service, the Service usually just checks the Token is valid and checks to see if there is a op sepcified Role on the Project they are authenticated to, and then the Service associates anything created/listed/etc with that Project.

In k8s, Projects COULD be mapped to namespaces, and in that mode, would function just like all other OpenStack services. Very similar to how Nova works. Nova Instances are namespaced at a Project level. Users switch namespaces by switching Project and getting a new scoped token.

Since k8s is not multitenant aware at present, kubernetes as as service (k8saas) is provided as part of Magnum. This allows a Project to create a few VM's, and launch k8s on it. In this type of k8s deployment, it is most likely intended to be used by Users of the Project that owns the VM's its launched on. AKA single tenant mode. In this case, Users should just be restrected to Users that have a specified Role on a specified Project so that only Users of that Project can perform actions. Keystone Authentication should still be used so that a seperate authentication mechanism shouldn't need to be used by the users. The exiting Keystone creds should be sufficient.

Does this explenation help clarify things? This is why its hard to capture all of this stuff in username/password auth too. Tokens do a lot of things.

This is also kind of why the authn and authz code in k8s is kind of blured with respect to Keystone, since Keystone itself does way way more then just authn. Its mostly about authz.